sled-agent/src/common/instance.rs

```diff
@@ -256,56 +279,82 @@ impl InstanceStates {
         // Update the instance record to reflect the result of any completed
         // migration.
         match observed.migration_status {
-            ObservedMigrationStatus::Succeeded => match self.propolis_role() {
-                // This is a successful migration out. Point the instance to the
-                // target VMM, but don't clear migration IDs; let the target do
-                // that so that the instance will continue to appear to be
-                // migrating until it is safe to migrate again.
-                PropolisRole::Active => {
-                    self.switch_propolis_id_to_target(observed.time);
-
-                    assert_eq!(self.propolis_role(), PropolisRole::Retired);
+            ObservedMigrationStatus::Succeeded => {
+                if let Some(ref mut migration) = self.migration {
+                    migration.state = MigrationState::Completed;
+                    migration.gen = migration.gen.next();
+                    migration.time_updated = observed.time;
+                } else {
+                    // XXX(eliza): we don't have an active migration state, but
+                    // we saw a migration succeed. this is almost certainly a
```

Comment:
Ah, yeah, I think you're right that we shouldn't be able to reach this case as a
result of Propolis misbehaving. It should probably be a debug assert (or perhaps
we should just `expect` it to be `Some`, honestly...)

Reply:
Huh..turns out we were wrong about this, at least in the simulated sled-agent: https://buildomat.eng.oxide.computer/wg/0/details/01J025F2Z61DTCC4HXS6MFWZXB/a2v6jXzafrXSXZxlVhUaz7pMhGxO8yGnPcmi9sRlIIWYqTay/01J025FBZT929NBEJKMF0JFY5J#S4656

That seems like it could be a test bug though, will investigate further.

Reply:
Oh, yeah, that's a test bug, these are unit tests that never call into the code that would create the `MigrationState`